keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1550 @@
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We extend the CRI protocol to contain information about the QoS-class
+resource assignment of containers and pods.
+
+Pod-level and container-level QoS-class resources are completely independent
+resource types. E.g. specifying something in the pod-level request does not
+mean specifying a pod-level default for all containers of the pod.
+
+Currently we identify two types of container-level QoS-class resources (RDT and
+blockio) but the API changes will be generic so that it will serve other
+similar resources in the future. Currently there are no immediately enabled
+pod-level QoS-class resources but we see usage scenarios for those in the
+future (communicating the pod QoS class to the runtime and enabling pod-level
+cgroup controls for blockio).
+
+We also extend the CRI protocol to support updates of QoS-class resource
+assignment of running containers. We recognize that currently container
+runtimes lack the capability to update either of the two types of QoS-class
+resources we have identified (RDT and blockio). However, there is no technical
+limitation in that and we are planning to implement update support for them
+in the future.
+
+We implement pod annotations the initial mechanism for Kubernetes users to
+control QoS-class resource assignment. We define two container-level QoS-class
+resources that can be controlled via annotations, i.e. RDT and blockio.
+
+We introduce a feature gate that enables kubelet to interpret pod annotations
+for controlling the RDT and blockio class of containers.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a user I want to minimize the interference of other applications to my
+workload by assigning it to a class with exclusive cache allocation.
+
+#### Story 2
+
+As a user I want to make sure my low-priority, I/O-intensive background task
+will not disturb more important workloads running on the same node.
+
+#### Story 3
+
+As a cluster administrator I want to throttle I/O bandwidths of certain
+DaemonSets, and I want that exact throttling values depend on the SSD model in
+my heterogenous cluster.
+
+#### Story 4
+
+As a user I want to assign a low priority task into an (RDT) class that limits
+the available memory bandwidth.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+This is only the first step in getting QoS-class resources supported in
+Kubernetes. Important pieces like resource assignment via pod spec, resource
```

Comment:
has this changed?

Reply:
Err, yes it has changed in the sense that we have better understanding what the complete solution would/could look like. I also have a working POC with all the future steps implemented. However, there are alternative ways to solve the future steps and not all details have been spelled out. I reworded this section and also moved non-goals that were really future steps to Goals section. WDYT?